tendermint/src/merkle.rs

```diff
@@ -61,88 +61,31 @@ pub trait MerkleHash {
             },
         }
     }
-}
-
-// A helper to copy GenericArray into the human-friendly Hash type.
-fn copy_to_hash(output: impl AsRef<[u8]>) -> Hash {
-    let mut hash_bytes = [0u8; HASH_SIZE];
-    hash_bytes.copy_from_slice(output.as_ref());
-    hash_bytes
-}
-
-impl<H> MerkleHash for H
-where
-    H: Digest<OutputSize = U32> + FixedOutputReset,
-{
-    fn empty_hash(&mut self) -> Hash {
-        // Get the output of an empty digest state.
-        let digest = self.finalize_reset();
-        copy_to_hash(digest)
-    }
-
-    fn leaf_hash(&mut self, bytes: &[u8]) -> Hash {
-        // Feed the data to the hasher, prepended with 0x00.
-        Digest::update(self, [0x00]);
-        Digest::update(self, bytes);
-
-        // Finalize the digest, reset the hasher state.
-        let digest = self.finalize_reset();
-
-        copy_to_hash(digest)
-    }
 
-    fn inner_hash(&mut self, left: Hash, right: Hash) -> Hash {
-        // Feed the data to the hasher: 0x1, then left and right data.
-        Digest::update(self, [0x01]);
-        Digest::update(self, left);
-        Digest::update(self, right);
-
-        // Finalize the digest, reset the hasher state
-        let digest = self.finalize_reset();
-
-        copy_to_hash(digest)
+    /// Compute a simple Merkle root from vectors of arbitrary byte vectors.
+    /// The leaves of the tree are the bytes of the given byte vectors in
+    /// the given order.
+    fn simple_hash_from_byte_vectors(&mut self, byte_vecs: &[Vec<u8>]) -> Hash {
```

Comment:
Why does this need to be converted to a method?
With the freestanding function, we can't misuse the API by e.g. calling it on a digest object with non-reset digest state.
`simple_hash_from_byte_vectors` is a one-stop-shop function that creates the digest state internally, hashes the input and returns the result.

Reply:
Likewise, I had looked at it from the opposite angle, but with the same concern. Having a bare function, so Merkle hasher could be triggered by a fake similar one. I encapsulated it to operate only on that data. btw, could be investigate more or even leave it untouched for now. alright